### PR TITLE
Add multi-turn support in qna file format

### LINKED ISSUE #448: Support for multi-turn QnA in .qna file format
- Multi-turn support
    - [x] Support **Follow up prompts:** as a list of markdown links [displaytext](#<question>) to other questions.
    - [x] Support ability to assign an ID to a QnA section and use that in Follow up prompts via <a id=”xxx”></a> and use this in a link via [displaytext](#xxx)
    - [x] Support ability to specify context only follow up prompts via `context-only`
- [x] Support filters to appear either before or after answer
- [x] Support ability to import all utterances from an .lu file, from specific intents
- [x] Do not require `markdown` for answer
- [x] Support > @kb.source = in comment for every QnA pair

Here’s a sample .qna document - 
```markdown
> source: custom editorial
# ?book flight
- can you help me book a flight?
```markdown
Sure. I can help book a flight.
\```

**Follow up prompts:**
- [seattle](#?seattle) `context-only`
- [redmond](#?redmond)
   - [joke](#?joke)

# ?seattle
```markdown
sure. I can help you book flights to seattle.
\```

# ?redmond
```markdown
Sorry, I do not fly to redmond.
\```
**Follow up prompts:**
- [joke](#?joke)

# ?Book cab
```markdown
you bet i can help with that
\```
**Follow up prompts:**
- [joke](#?joke)

# ?joke
**Filters:**
- propertyName = b
```markdown
ha ha ha
\```

# ?joke
**Filters:**
- propertyName = a
```markdown
I dont know!
\```
```

### PR DESCRIPTION
**Pull request by vishwacsena**
Fixes #448

- [x] Multi-turn support
    - [x] Support **Follow up prompts:** as a list of markdown links [displaytext](#<question>) to other questions.
    - [x] Support ability to assign an ID to a QnA section and use that in Follow up prompts via <a id=”xxx”></a> and use this in a link via [displaytext](#xxx)
    - [x] Support ability to specify context only follow up prompts via `context-only`
- [x] Support filters to appear either before or after answer
- [x] Support ability to import all utterances from an .lu file, from specific intents
- [x] Do not require `markdown` for answer
- [x] Support > @kb.source = in comment for every QnA pair

### CODE REVIEW

**Comment by vishwacsena** [comment]
Fixes #448

File: packages/lu/src/parser/lu/luMerger.js [hunk 1]
```diff
@@ -141,6 +142,48 @@ const haveLUISContent = function(blob) {
     (blob.composites.length > 0));
 }
 
+const resolveReferencesInQuestions = async function(allParsedContent) {
+    (allParsedContent.QnAContent || []).forEach(qnaModel => {
+        (qnaModel.qnaJsonStructure.qnaList || []).forEach(qaPair => {
+            let refQIdx = [];
+            (qaPair.questions || []).forEach((question, idx) => {
+                // Is this question a reference? 
+                if (helpers.isUtteranceLinkRef(question || '')) {
```
Comment (munozemilio) [inline packages/lu/src/parser/lu/luMerger.js]: nit. we could switch to a normal for and continue if this conditions is not met.

File: packages/lu/src/parser/lufile/qnaSection.js [hunk 2]
```diff
@@ -18,7 +20,64 @@ class QnaSection {
...
+            return idTextParsed.groups.idCaptured || undefined;
+        }
+        return undefined;
+    }
+
+    ExtractPrompts(parseTree) {
+        let promptDefinitions = [];
+        let errors = [];
+        let promptSection = parseTree.qnaDefinition().promptSection();
+        if (promptSection) {
```
Comment (munozemilio) [inline packages/lu/src/parser/lufile/qnaSection.js]: nit return if null or undefined to remove one level of nested code.
```
if(!promptSection){
return { promptDefinitions, errors};
}
```

File: packages/lu/src/parser/qna/qnamaker/qnamaker.js [hunk 3]
```diff
@@ -45,11 +46,98 @@ class QnAMaker {
...
+        resolveMultiTurnReferences(this.qnaList);
+        resolveQnAIds(this.qnaList);
     }
 }
 
 module.exports = QnAMaker
 
+
+const resolveMultiTurnReferences = function(qnaList) {
+    let qnaPairsWithMultiTurn = qnaList.filter(item => item.context.prompts.length !== 0);
```
Comment (munozemilio) [inline packages/lu/src/parser/qna/qnamaker/qnamaker.js]: will item.context.prompts always exist?

File: packages/lu/src/parser/lufile/qnaSection.js [hunk 4]
```diff
@@ -73,9 +134,10 @@ class QnaSection {
 
     ExtractAnswer(parseTree) {
         let multiLineAnswer = parseTree.qnaDefinition().qnaAnswerBody().multiLineAnswer().getText().trim();
-        let answer = multiLineAnswer.slice(11, multiLineAnswer.length - 3);
-        
-        return answer;
+        // trim first and last line
+        var answerRegexp = /^```(markdown)?\n?(?<answer>(.|\n)*)```$/gim;
```
Comment (munozemilio) [inline packages/lu/src/parser/lufile/qnaSection.js]: nit. let

File: packages/lu/src/parser/lufile/qnaSection.js (hunk 2 above)
Comment (vishwacsena) [reply packages/lu/src/parser/lufile/qnaSection.js]: Fixed.

File: packages/lu/src/parser/lufile/qnaSection.js (hunk 4 above)
Comment (vishwacsena) [reply packages/lu/src/parser/lufile/qnaSection.js]: fixed.

File: packages/lu/src/parser/lu/luMerger.js (hunk 1 above)
Comment (vishwacsena) [reply packages/lu/src/parser/lu/luMerger.js]: Keeping this as is.

**Pull request merged**